### DISCUSSION TITLE
enabling tcp keepalive on WebSockets?

### DISCUSSION THREAD

**Discussion by rgillan** [oven-sh/bun#14344] [Q&A]
Just wondering if there was a way we can activate tcp keepalive on socket connections.  In nodejs we can simply:
```socket.setKeepAlive(true, 10000);```
but not sure if anything like this is possible with Bun.serve.

**Comment by guest271314** [comment]
Are you sure Bun's WebSocket implementation closes the connection after a certain amount of time now?

**Reply by rgillan** [reply]
Actually re-running those tests next week and will report here.  This question is more about the low level tcp keepalive which helps when there's underlying networking impairments and one way failures.  We also run WebSocket keep alive at the app layer.

**Reply by guest271314** [reply]
The last time I checked TCP connections do not have a timeout.

**Reply by rgillan** [reply]
https://en.wikipedia.org/wiki/Keepalive

**Reply by guest271314** [reply]
None of these TCP servers time out https://github.com/guest271314/telnet-client/tree/user-defined-tcpsocket-controller-web-api/direct-sockets. This WebSocket and HTTP server https://github.com/guest271314/direct-sockets-http-ws-server/blob/main/assets/script.js run from Chrome browser using `TCPServerSocket` with `WebSocketStream` used in the browser https://github.com/guest271314/direct-sockets-http-ws-server/blob/main/examples/ws.js doesn't time out.

**Reply by guest271314** [reply]
Here's the JavaScript runtime agnostic WebSocket server standalone https://gist.github.com/guest271314/735377527389f1de6145f0ac71ca1e86.

**Reply by rgillan** [reply]
sorry but I'm not sure what you're asking/wanting here.  This discussion is asking how we can turn on tcp keepalive within Bun as we can in nodejs.  If you have other topics then can I suggest you create one specifically for what you are asking?

**Reply by guest271314** [reply]
What I'm saying is I have not observed Bun's TCP implementation timing out at all. Test and verify for yourself.

**Reply by Jarred-Sumner** [reply]
@guest271314 TCP keep-alive is a socket option in TCP, and is (confusingly) unrelated to HTTP keep-alive

**Reply by guest271314** [reply]
Right. I have not observed the Bun TCP socker server implementation timing out at all, as of my last tests. https://github.com/guest271314/telnet-client/blob/user-defined-tcpsocket-controller-web-api/direct-sockets/bun_echo_tcp.js. Test and verify.

**Reply by guest271314** [reply]
Once the TCP server is started you can send HTTP, HTTP/1.1 communication over that TCP connection. TCP protocol doesn't care. You just have to parse the respective headers.

**Reply by guest271314** [reply]
Deno's built-in WebSocket server at one point did time out. Deno folks fixed that though.

**Reply by guest271314** [reply]
> @guest271314 TCP keep-alive is a socket option in TCP, and is (confusingly) unrelated to HTTP keep-alive

I'm not seeing Bun 1.1.31 built-in WebSocket server time out at all on Linux.

`bun-ws.js`

```
Bun.serve({
  port: 3000,
  fetch(req, server) {
    // upgrade the request to a WebSocket
    if (server.upgrade(req)) {
      return; // do not return a Response
    }
    return;
  },
  websocket: {
    open(ws) {
      console.log("open");
    },
    message(ws, message) {
      console.log(message);
      ws.send(message);
    },
    close(ws) {
      console.log("close", ws);
    },
    error(ws, e) {
      console.log(e);
    },
    drain(ws) {
      console.log(drain);
    },
  },
});
```

In DevTools on Chromium 131 

```
var wss = new WebSocketStream("ws://0.0.0.0:3000");
console.log(wss);
wss.closed.catch((e) => {});
wss.opened.catch((e) => {});
var {
  readable,
  writable,
} = await wss.opened.catch(console.error);
var writer = writable.getWriter();
var abortable = new AbortController();
var {
  signal,
} = abortable;
// .pipeThrough(new TextDecoderStream())
var pipe = readable.pipeTo(new WritableStream({
  start(c) {
    console.log("Start", c);
  },
  async write(v) {
    console.log(v, decoder.decode(v));
  },
  close() {
    console.log("Socket closed");
  },
  abort(reason) { // console.log({ reason });
  },
}), {
  signal,
}, ).then(() => ({
  done: true,
  e: null
})).catch((e) => ({
  done: true,
  e
}));

var encoder = new TextEncoder();
var decoder = new TextDecoder();
var encode = (text) => encoder.encode(text);
await writer.write(encode("X"));
```

**Comment by Jarred-Sumner** [comment]
@rgillan can you file an issue and we will implement an option for this somehow

**Reply by rgillan** [reply]
@Jarred-Sumner will do
